File: typescript/src/pages/InventoryPage.ts
import { type Page, type Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage.js';

export type SortOption =
  | 'az'
  | 'za'
  | 'lohi' // price low to high
  | 'hilo'; // price high to low

/**
 * Page object for the products / inventory listing.
 */
export class InventoryPage extends BasePage {
  readonly title: Locator;
  readonly items: Locator;
  readonly cartLink: Locator;
  readonly cartBadge: Locator;
  readonly sortDropdown: Locator;

  constructor(page: Page) {
    super(page);
    this.title = page.locator('.title');
    this.items = page.locator('.inventory_item');
    this.cartLink = page.locator('.shopping_cart_link');
    this.cartBadge = page.locator('.shopping_cart_badge');
    this.sortDropdown = page.getByTestId('product-sort-container');
  }

  async expectLoaded(): Promise<void> {
    await expect(this.title).toHaveText('Products');
    await expect(this.items.first()).toBeVisible();
  }

  /** Scope an "Add to cart" button to the card with the given product name. */
  private addToCartButton(productName: string): Locator {
    return this.page
      .locator('.inventory_item')
      .filter({ hasText: productName })
      .getByRole('button', { name: 'Add to cart' });
  }

  async addItemToCart(productName: string): Promise<void> {
    await this.addToCartButton(productName).click();
  }

  async cartCount(): Promise<number> {
    if (!(await this.cartBadge.isVisible())) return 0;
    return Number(await this.cartBadge.innerText());
  }

  async expectCartCount(count: number): Promise<void> {
    if (count === 0) {
      await expect(this.cartBadge).toBeHidden();
    } else {
      await expect(this.cartBadge).toHaveText(String(count));
    }
  }

  async sortBy(option: SortOption): Promise<void> {
    await this.sortDropdown.selectOption(option);
  }

  async productNames(): Promise<string[]> {
    return this.page.locator('.inventory_item_name').allInnerTexts();
  }

  async openCart(): Promise<void> {
    await this.cartLink.click();
  }
}

File: python/pages/inventory_page.py
"""Page object for the products / inventory listing."""

from __future__ import annotations

from playwright.sync_api import Locator, Page, expect

from .base_page import BasePage


class InventoryPage(BasePage):
    def __init__(self, page: Page) -> None:
        super().__init__(page)
        self.title = page.locator(".title")
        self.items = page.locator(".inventory_item")
        self.cart_link = page.locator(".shopping_cart_link")
        self.cart_badge = page.locator(".shopping_cart_badge")
        self.sort_dropdown = page.get_by_test_id("product-sort-container")

    def expect_loaded(self) -> None:
        expect(self.title).to_have_text("Products")
        expect(self.items.first).to_be_visible()

    def _add_button(self, product_name: str) -> Locator:
        return (
            self.items.filter(has_text=product_name)
            .get_by_role("button", name="Add to cart")
        )

    def add_item_to_cart(self, product_name: str) -> None:
        self._add_button(product_name).click()

    def expect_cart_count(self, count: int) -> None:
        if count == 0:
            expect(self.cart_badge).to_be_hidden()
        else:
            expect(self.cart_badge).to_have_text(str(count))

    def sort_by(self, value: str) -> None:
        self.sort_dropdown.select_option(value)

    def product_names(self) -> list[str]:
        return self.page.locator(".inventory_item_name").all_inner_texts()

    def open_cart(self) -> None:
        self.cart_link.click()
